Feat: Scoord markers & markups.

igoroctaviano opened the pull request:
Add markers for scoords and markups.

## Measurement
![Screen Shot 2020-11-11 at 15 50 19](https://user-images.githubusercontent.com/13886933/98853322-b85f8100-2437-11eb-8d90-2e9a10550149.png)

- [x] Measurement (LineString / Polygon / Circle)
- [x] Arrow (Point / LineString) 
- [x] FreeText (Point)

hackermd commented:
Thanks @igoroctaviano! This looks great!!

Before we delve into the details of the implementation, I would like to discuss the general design. Specifically, concepts such as "length" or "free text" do not represent geometries in my opinion, but are rather measurements or qualitative evaluations of regions of interests, which are defined by geometries (`POLYGON`, `POINT`, etc.). I am wondering whether we could generalize this and add support for different types of ROI properties that map onto DICOM SR content item value types (`CODE` -> qualitative evaluations in form of well-defined terminologies, `TEXT` -> qualitative evaluations in form of "free text", `NUM` -> measurements such as "length", etc.).

We could for example use the `updateROI()` method to add properties to an existing ROI (). Each value type could have a predefined style associated with it that would define how the property gets rendered alongside the ROI. This could be automatically done upon drawing using the `ROI_DRAWN` event.

What do you think?

hackermd commented:
@igoroctaviano I added new methods to both the `ROI` and the `VolumeViewer` classes to add ROI Measurements and Qualitative Evaluations, following [TID 1410 Planar ROI Measurements and Qualitative Evaluations](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_1410)). These are available with release v0.21.1.

igoroctaviano commented:
> Thanks @igoroctaviano! This looks great!!
> 
> Before we delve into the details of the implementation, I would like to discuss the general design. Specifically, concepts such as "length" or "free text" do not represent geometries in my opinion, but are rather measurements or qualitative evaluations of regions of interests, which are defined by geometries (`POLYGON`, `POINT`, etc.). I am wondering whether we could generalize this and add support for different types of ROI properties that map onto DICOM SR content item value types (`CODE` -> qualitative evaluations in form of well-defined terminologies, `TEXT` -> qualitative evaluations in form of "free text", `NUM` -> measurements such as "length", etc.).
> 
> We could for example use the `updateROI()` method to add properties to an existing ROI (). Each value type could have a predefined style associated with it that would define how the property gets rendered alongside the ROI. This could be automatically done upon drawing using the `ROI_DRAWN` event.
> 
> What do you think?

Hi @hackermd thank you for the feedback. That's right, they represent exactly what you just described. I’ll update the terms used here for sure.

So, I'm not super experienced with SR yet but does it support storing measurements/evaluations/labels and how they will be displayed? I mean, I thought first that this metadata would be the case for DICOM presentation state? 

About the SR content items, for now, beyond roi/scoord, I’m storing just the type (arrow, length, free text) by adding it during hydration as part of the tracking Identifier using the [dcmjs.sr.templates.PlanarROIMeasurementsAndQualitativeEvaluations](https://github.com/dcmjs-org/dcmjs/blob/b39c0029042a0044e69e57b3df5b6c6aa68b1015/src/sr/templates.js#L1095) template which basically means "any reasonable temp place to quickly set up a stow demo" using these annotations. I still need to do some work around how these annotations should actually be stored, maybe you can shed some light on it too? that would be great.

hackermd commented:
> So, I'm not super experienced with SR yet but does it support storing measurements/evaluations/labels and how they will be displayed? I mean, I thought first that this metadata would be the case for DICOM presentation state?

DICOM SR allows for communication of measurements and evaluations of regions of interest (ROIs). Here is a mapping of GeoJSON (used by Openlayers) to DICOM terminology:

* `feature` -> measurement (continuous, `NUM` value type) or qualitative evaluation (categorical, `CODE` or `TEXT` value type)
* `goemetry` -> region (`SCOORD3D` value type)

TID 1410 Planar ROI Measurements and Qualitative Evaluations is extensible, so in principle you can store all kinds of information along with a ROI, as long as it can be represented with one of the value types (`NUM`, `CODE`, `TEXT`, ...).

Note, however, that DICOM SR is not intended for communication of representation states to exactly reproduce how regions or measurements/evaluations derived thereof should be displayed. However, that is not that important for interoperability in my opinion and can be handled differently by each display system as long as it doesn't alter the semantics.

Arrows are particularly problematic in this regard, since their meaning can be ambiguous. I think its perfectly fine to use an arrow to represent a point, as long as it's clear that the length or orientation of the arrow doesn't convey any meaning. In other words, an arrow is just a specific representation of a point.

I am in principle fine with using the term `CustomGeometry` internally, as long as this remains an implementation detail. However, I would not expose concepts like arrows via the API as a separate `geometryType`. I would prefer to call them "markers" instead of "geometries" even if they are derived from the Openlayers `Geometry` class.

What about providing users an option to configure how different types of `Scoord3d` should be represented using different markers? For example, whether a point should be represented as a dot marker, a map marker, or an arrow marker.

In addition, I think we could generalize the concept of "length", which represents a line measurement to different `Scoord3d` as well as different measurements and qualitative evaluations. For example, we could similarly show the area measurement of a `Polygon` or the [finding](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_dfd59901-9eaf-4594-86b4-d9df83d9972f) of any `Scoord3d`. We should be able to render every measurement (`NUM` value type) and qualitative evaluation (`CODE` or `TEXT` value type) along with their `Scoord3d`.

Let me know what you think. In the meantime, I will take a closer look at your implementation and start a review.

hackermd commented:
> About the SR content items, for now, beyond roi/scoord, I’m storing just the type (arrow, length, free text) by adding it during hydration as part of the tracking Identifier using the dcmjs.sr.templates.PlanarROIMeasurementsAndQualitativeEvaluations template which basically means "any reasonable temp place to quickly set up a stow demo" using these annotations. I still need to do some work around how these annotations should actually be stored, maybe you can shed some light on it too? that would be great.

In DICOM SR, concepts are represented as name-value pairs. However, "length", "arrow" and "free text" represent very different concepts. A "length" is a measurement of a region defined by a `POLYLINE`, while an "arrow" is a representation of a region defined by a `POINT` and "free text" is basically a comment

* Length:
  - [Image region](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_208032d8-b7e8-4af0-bd71-840b23278e24): `ValueType` = `SCOORD3D`, `GraphicType` = `POLYLINE`
  - [Length](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7470.html#para_e84a5c40-73ad-4959-9b27-6a2f95250a71) (in [TID1419 ROI Measurements](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_1419)): `ValueType` = `NUM`
* Arrow:
  - [Image region](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_208032d8-b7e8-4af0-bd71-840b23278e24): `ValueType` = `SCOORD3D`, `GrahicType` = `POINT`
  - ???
* Free text:
  - [Image region](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_208032d8-b7e8-4af0-bd71-840b23278e24): `ValueType` = `SCOORD3D`, any `GraphicType`
  - [Comment](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_D.html#DCM_121106): `ValueType` = `TEXT`

igoroctaviano commented:
> > So, I'm not super experienced with SR yet but does it support storing measurements/evaluations/labels and how they will be displayed? I mean, I thought first that this metadata would be the case for DICOM presentation state?
> 
> DICOM SR allows for communication of measurements and evaluations of regions of interest (ROIs). Here is a mapping of GeoJSON (used by Openlayers) to DICOM terminology:
> 
> * `feature` -> measurement (continuous, `NUM` value type) or qualitative evaluation (categorical, `CODE` or `TEXT` value type)
> * `goemetry` -> region (`SCOORD3D` value type)
> 
> TID 1410 Planar ROI Measurements and Qualitative Evaluations is extensible, so in principle you can store all kinds of information along with a ROI, as long as it can be represented with one of the value types (`NUM`, `CODE`, `TEXT`, ...).
> 
> Note, however, that DICOM SR is not intended for communication of representation states to exactly reproduce how regions or measurements/evaluations derived thereof should be displayed. However, that is not that important for interoperability in my opinion and can be handled differently by each display system as long as it doesn't alter the semantics.
> 
> Arrows are particularly problematic in this regard, since their meaning can be ambiguous. I think its perfectly fine to use an arrow to represent a point, as long as it's clear that the length or orientation of the arrow doesn't convey any meaning. In other words, an arrow is just a specific representation of a point.
> 
> I am in principle fine with using the term `CustomGeometry` internally, as long as this remains an implementation detail. However, I would not expose concepts like arrows via the API as a separate `geometryType`. I would prefer to call them "markers" instead of "geometries" even if they are derived from the Openlayers `Geometry` class.
> 
> What about providing users an option to configure how different types of `Scoord3d` should be represented using different markers? For example, whether a point should be represented as a dot marker, a map marker, or an arrow marker.
> 
> In addition, I think we could generalize the concept of "length", which represents a line measurement to different `Scoord3d` as well as different measurements and qualitative evaluations. For example, we could similarly show the area measurement of a `Polygon` or the [finding](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#para_dfd59901-9eaf-4594-86b4-d9df83d9972f) of any `Scoord3d`. We should be able to render every measurement (`NUM` value type) and qualitative evaluation (`CODE` or `TEXT` value type) along with their `Scoord3d`.
> 
> Let me know what you think. In the meantime, I will take a closer look at your implementation and start a review.

Well, by providing an option so the user can configure how different types of scoord should be represented using different markers you mean keeping this logic on the user's repo and adding enough config so the user could update the way a scoord is rendered? if that's the case, we would not reuse the logic along with the library.

If thats not the case, another approach: 
- I would update the ROI class so it would have a **roiType** property which could have one of the values: length, arrow, freetext that would be used by the addROI function to create the appropriate styled version of Point/Polygon/Polyline feature for that specific ROI and a marker property that would look like:
```
    marker: { 
      valueKey: ROI properties key that will be used to key off the value of the marker,
      displayLink: some other marker related properties
    }
```
In this case, I would add a new option to activateDrawInteraction where I would pass along with the geometryType, a roiType property that would replace the geometryName I'm currently using.

What do you think? let me know if everything feels too weird, there's a lot of ways of doing it 😄

igoroctaviano commented:
@hackermd I have updated the code with the following changes:
- CustomGeometry became Markers, now the option marker will be used to determine the new representation of a Point/etc.
- Length became Measurement and now it can be activated for Polygons and Circles (area).
- Arrow now can be used for lines and points.

Example of usage: pass marker option to activateDrawInteraction.
```
activateDrawInteraction({ ...toolOptions, marker: 'arrow' });
```
I have added some examples adding measurements and evaluations, I need your help to get these right. 

Can you review the code the way its now? i need to move this forward, let me know if you're busy and if there's anyone I can reach out to review.

hackermd reviewed (changes requested):
@igoroctaviano thank you for these contributions. This is really nice code and I think the functionality will be extremely useful.

I have added a couple of major comments regarding overall API design and aspects related to the DICOM standard. In addition, I have already added a few minor comments related to implementation details and coding style. However, before delving into the implementation details, I would like to first discuss the overall design with you:

1. Distinction between markers (graphics -> ROI spatial coordinates) and markup (annotations -> ROI measurements and qualitative evaluations):
    - **Markers (graphics)**
        + DICOM supports the following [Graphic Types](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html#sect_C.18.9.1.2) for Content Items of value type [3D spatial coordinates (SCOORD3D)](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.9.html) to define a region of interest (ROI): `POINT`, `MULTIPOINT`, `POLYLINE`, `POLYGON`, `ELLIPSE` and `ELLIPSOID`. The dicom-microscopy-viewer library implements those vector graphics as [Scoord3D](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/blob/4bf573dd9a85f4a106e8e06718c34ace7e98bca5/src/scoord3d.js). They are rendered using openlayers [GeometryType](https://openlayers.org/en/latest/apidoc/module-ol_geom_GeometryType.html).
        + To me, a "marker" represents a specific representation (rendering) of a SCOORD3D with a given Graphic Type. For example, a `POINT` may be rendered as a single point or as an arrow (with the tip of the arrow pointing to the coordinate defined by the [Point](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/blob/4bf573dd9a85f4a106e8e06718c34ace7e98bca5/src/scoord3d.js#L72)).
       + We need to clearly distinguish between the SCOORD3D, which defines a ROI as a set of spatial coordinates, and things like arrows, which represent a specific rendering of the ROI. The latter are essentially styling and they must not convey any semantic meaning.
    - **Markup (annotations)** 
        + DICOM supports different kinds of [ROI measurements and qualitative evaluations](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_A.html#sect_TID_1410) using Content Items of value type [numeric measurements (NUM)](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.html#sect_C.18.1), free text (TEXT), or [code (CODE)](http://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.18.2.html).
        + The dicom-microscopy-viewer currently stores measurements and evaluations on the `properties` attribute of a openlayers [Feature](https://openlayers.org/en/latest/apidoc/module-ol_Feature-Feature.html).
        + To me, a "markup" represents either a numeric measurement (NUM) or a qualitative evaluation (CODE or TEXT). These can be added to a [ROI](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/blob/4bf573dd9a85f4a106e8e06718c34ace7e98bca5/src/roi.js#L45).
    - The viewer API exposes ROI measurements and evaluations via the dicom-microscpopy-viewer `ROI`, but handles them under the hood via the openlayers `Feature`. Similarly the dicom-microscopy-viewer `Scoord3D` is handled via openlayers `Geometry` and derived measurements and evaluations are handled via openlayers `properties`. The public API should only expose dicom-microscopy-viewer `ROI` objects, but the underlying low-level implementations concerned with rendering vector graphics should probably work with openlayers `Feature` directly to avoid frequent conversions (different coordinate systems, etc.).
    - I would make a clear distinction between **Markers** and **Markup**, such that the former only deals with the vector graphics (spatial coordinates -> openlayers `Feature.geometry`), while the latter deals with annotations derived thereof (measurements and evaluations -> openlayers `Feature.properties`).
    - So far, the viewer didn't render measurements or evaluations and I think the "marker" definitions are a great abstraction layer for these concepts. However, as mentioned above, I think we should separate "marker" definitions (openlayers `Feature.geometry`) from "markup" definition (openlayers `Feature.properties`). They can then be handled via a manager class (e.g., `_AnnotationManager`), which links markers and markup using openlayers `Feature`.
    - I think we should handle measurements and evaluations using SR Content Items [NumContentItem](https://github.com/dcmjs-org/dcmjs/blob/678543c46df63ec9c11e1f5a97e1776242fea2a6/src/sr/valueTypes.js#L302), [TextContentItem](https://github.com/dcmjs-org/dcmjs/blob/678543c46df63ec9c11e1f5a97e1776242fea2a6/src/sr/valueTypes.js#L166) and [CodeContentItem](https://github.com/dcmjs-org/dcmjs/blob/678543c46df63ec9c11e1f5a97e1776242fea2a6/src/sr/valueTypes.js#L149) and also store them on the openlayers `Feature.properties` object. Free text is straight forward to handle, and can be provided by users via a text input field. However, for `NUM` and `CODE` we need to keep track of additional information, including the unit of measurement, the code value, and the coding scheme designator. Measurements are probably only created by applications and I cannot think of a use case were users would provide the values directly. Coded evaluations should be based on defined value sets (DICOM calls those [Context Groups](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_7.html)) and values should by provided by users only through a drop down menu (or something similar).
    - We may want to add an abstraction layer around those concepts (e.g., `addMeasurment()` and `addEvaluation()`) and include runtime type checks. I also have created Typescript type definitions. I haven't added them in the repo yet, but will work on this within the next weeks.

2. Styling
    - The viewer API allows the developer to specify the style of ROIs, including the stroke and fill color of rendered vector graphics. Currently, the style for markers and markups is hardcoded. We should propagate the style properties set via the viewer API to the underlying rendering utilities, such that graphics are rendered consistenly.

3. Emphasize that the implementations of markers and markup are private and should not be exposed via the public API. At a minimum, I would add a leading underscore to the class names. For the implementation of the viewer, I further use [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) to hide implementation details and prevent access to the underlying openlayers API.

hackermd reviewed (changes requested):
It is coming together nicely! I just added one comment related to the coding of `NUM` measurements. However, the comment applies similarly to `TEXT` and `CODE` evaluations.

In order to store or retrieve annotations in DICOM SR documents, we need to code the name of the annotated concept. In case of `NUM` or `CODE` we also need to code the unit or value, respectively. 

In case of `NUM`, we know what measurement we are taking. Therefore, we don't need any user input, but need to pass this information to the application.

In case of `TEXT` and `CODE`, the user (or the application) needs to specify what is being (or supposed to be) annotated. The application thus needs to choose a code from a context group (value set, enum) and pass this information to the library, such that the library can store and track it on the created `ROI`.

We should be able to handle all this this via the existing events (`EVENT.ROI_ADDED`, `EVENT.ROI_DRAWN`, `EVENT.ROI_SELECTED`, etc.) and exchange the information across the application-library interface in form of `ROI` objects. However, we need to make this that this is probably wired with the `_AnnotationManager`, such that everything stays in sync. The viewer will need to translate/transform the dicom-microscopy-viewer `ROI` objects to openlayes `Feature` objects and vice versa.

igoroctaviano commented:
Hey Markus, thanks again for the review.

So, just to let you know, I'm not 100% allocated on this but I'm gradually
updating the code locally, I had pushed some of it recently but there's
more to go.
I'll try to push today or tomorrow after I cover the last comment. And I
think after that we can schedule another meeting to align things? let me
know then I'll send an email to Ellen.

*--*

*Best regards,*


*Igor Octaviano*


Skype: igoroctaviano
Blog/website: igoroctaviano.com
Linkedin: linkedin.com/in/igoroctaviano/

Esta mensagem contém informações confidenciais e/ou privilegiadas. Se você
não for o destinatário ou a pessoa autorizada a receber esta mensagem, não
deve usar, copiar ou divulgar as informações nela contidas ou tomar
qualquer ação baseada nessas informações. Esse ambiente é monitorado.


On Wed, Jan 27, 2021 at 9:07 PM Markus D. Herrmann <notifications@github.com>
wrote:

> *@hackermd* requested changes on this pull request.
>
> It is coming together nicely! I just added one comment related to the
> coding of NUM measurements. However, the comment applies similarly to TEXT
> and CODE evaluations.
>
> In order to store or retrieve annotations in DICOM SR documents, we need
> to code the name of the annotated concept. In case of NUM or CODE we also
> need to code the unit or value, respectively.
>
> In case of NUM, we know what measurement we are taking. Therefore, we
> don't need any user input, but need to pass this information to the
> application.
>
> In case of TEXT and CODE, the user (or the application) needs to specify
> what is being (or supposed to be) annotated. The application thus needs to
> choose a code from a context group (value set, enum) and pass this
> information to the library, such that the library can store and track it on
> the created ROI.
>
> We should be able to handle all this this via the existing events (
> EVENT.ROI_ADDED, EVENT.ROI_DRAWN, EVENT.ROI_SELECTED, etc.) and exchange
> the information across the application-library interface in form of ROI
> objects. However, we need to make this that this is probably wired with the
> _AnnotationManager, such that everything stays in sync. The viewer will
> need to translate/transform the dicom-microscopy-viewer ROI objects to
> openlayes Feature objects and vice versa.
> ------------------------------
>
> In src/annotations/markups/utils/getUnitsSuffix.js
> <https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43#discussion_r565712292>
> :
>
> > @@ -0,0 +1,42 @@
> +import { getPointResolution } from 'ol/proj';
> +
> +const UnitsEnum = { METERS: 'm' };
> +const DEFAULT_DPI = 25.4 / 0.28;
> +
> +const getUnitsSuffix = (view) => {
>
> For units we will ultimately need a mapping form openlayers units to DICOM units
> of measurements
> <http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_7.2.2.html>,
> which are based on the Unified Units of Measure (UCUM)
> <https://en.wikipedia.org/wiki/Unified_Code_for_Units_of_Measure>
> standard. For planar size measurements:
>
>    - CID 7470 Linear Measurements
>    <http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7470.html>
>    and CID 7460 Units of Linear Measurement
>    <http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7460.html>
>    - CID 7471 Area Measurements
>    <http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7471.html>
>    and CID 7461 Units of Area Measurement
>    <http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7461.html>
>
> ------------------------------
>
> In src/annotations/markups/utils/getUnitsSuffix.js
> <https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43#discussion_r565717915>
> :
>
> > @@ -0,0 +1,42 @@
> +import { getPointResolution } from 'ol/proj';
> +
> +const UnitsEnum = { METERS: 'm' };
> +const DEFAULT_DPI = 25.4 / 0.28;
> +
> +const getUnitsSuffix = (view) => {
>
> We will need to specify the concepts that are being annotated (i.e., the
> kind of markup) and their unit, such that we can later store them in a
> DICOM SR document. For example, the length of a line could be coded as
> follows:
>
>    - name: ["410668003", "SCT", "Length"]
>    - unit: ["mm", "UCUM", "millimeter"]
>    - value: 2.3
>
> I think it would be best to store and pass the information around as
> instances of dcmjs.sr.valueTypes.NumContentItem (or a more compact
> representation).
>
> The markup manager can then render this for display to the user using the
> "Meaning" of the name and unit codes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43#pullrequestreview-577862684>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADJ6LVPMXHEUETCWFAPUYQ3S4CTC7ANCNFSM4TZEHOUQ>
> .
>

igoroctaviano commented on src/annotations/markups/utils/getUnitsSuffix.js:
```diff
@@ -0,0 +1,42 @@
+const UnitsEnum = { METERS: 'm' };
+const DEFAULT_DPI = 25.4 / 0.28;
+
+const getUnitsSuffix = (view) => {
```
@hackermd I have pushed some code that keeps the ROI's measurements and evaluations generation in sync with other custom feature properties like 'label' which allows the client app to update the label of the freetext. 'area', 'length' which gets updated every change in the geometry, and 'marker' / 'markup' which defines the style and behavior of markers and markups. I think it's simpler to work with the feature properties directly which will update the corresponding ROI object. You can take a look at the AnnotationManager where all this logic lies. 

I'm not using CODE evaluations for now because our client app only required freetext working but it's not hard to do, we just need to set another property like CODE that would be stored in a separate code content item and updated like the label property.

Using OHIF, I could store and retrieve the SR as expected for measurements and freetext. Now I'm not sure how I'm going to store information related to the marker itself since this is just a presentational state. Maybe another text content item could store this info (arrow marker) that would work only for this specific application? (not ideal)

Just let me know if you prefer to schedule a meeting to show you.

igoroctaviano commented on src/markers/MarkerManager.js:
```diff
@@ -0,0 +1,427 @@
+    const styleTag = document.createElement('style');
+    styleTag.innerHTML = `
+      .ol-tooltip {
+        color: #9ccef9;
```
This was updated in recent changes.

hackermd replied:
Are you sure? It looks like the default color is used. When the style of a roi gets set, it will not be applied to the text of measurement markups. It's a minor thing, we can also improve in a future version.

igoroctaviano replied:
Maybe it's something on my side then. I'm always changing and testing with my OHIF fork which sets highlighting colors when hovering the annotations. I'll take a look.

hackermd replied:
Thanks. This is non-blocking. In fact, it may be better to not color the markup text along with the marker. Instead, we could put a background box behind the text to ensure the text can be read independent of image content (staining, etc.). What do you think?

hackermd commented on src/annotations/markups/measurement.js:
```diff
@@ -0,0 +1,77 @@
+const isMeasurement = (feature) =>
+  Enums.Markup.Measurement === feature.get("markup");
+
+const MeasurementMarkup = (api) => {
```
What exactly is the `api` parameter? I personally find it difficult to understand what `api` represents and what name-pairs the object contains.

igoroctaviano replied:
Updated to explicit params, thanks

hackermd commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+    /** Markers */
+    this[Marker.Arrow] = ArrowMarker(this.props);
+
+    this.props.markupManager = new _MarkupManager({
```
I have trouble understanding what's going on here. It's confusing that `this.props` gets passed to the constructor of the markup and marker classes without a `markupManager` property, but these classes internally access the `markupManager` property on the passed options object (`api`). See also above [comment](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43/files#r584433586).

Can this be refactored in a way that is more intuitive?

igoroctaviano replied:
Updated this logic, thanks

hackermd reviewed (changes requested):
The PR contains lots of goodies!! Thank you for addressing my previous comments. Most of my remaining comments are about implementation details and related to standard compliance. However, I have a couple of questions regarding the interface (signature of functions/methods) that we should discuss.

The PR includes many new features and I have yet to test all of them manually. We should also start thinking about adding unit tests to cover the new features.

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+   */
+  getContentItemCodeMeaning(contentItem) {
+    const { ConceptNameCodeSequence } = contentItem;
+    return ConceptNameCodeSequence.length
+      ? ConceptNameCodeSequence[0].CodeMeaning
+      : ConceptNameCodeSequence.CodeMeaning;
```
Resolved

hackermd replied:
Has this changed? Am I missing something?

hackermd replied:
Somehow the `Files changed` tab shows me outdated file contents..

igoroctaviano commented on src/enums.js:
```diff
@@ -0,0 +1,47 @@
+  FreeTextEvaluation: "freetext",
+};
+
+export const FeatureEvents = {
```
We just have one enum file, that's why they are here but we can create different enum files, that's what you meant?

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+
+    const index = measurements.findIndex((measurement) => {
+      const meaning = this.getContentItemCodeMeaning(measurement);
+      return meaning === codeMeaning;
```
Updated, now I'm using equals or fallback to a built-in version of the equals method

hackermd replied:
My main concern here is that the comparison is performed based on the value of the *CodeMeaning* attribute. That is not a good idea and a combination of *CodeValue* and *CodingSchemeDesignator* should be used instead. Take a look at how we implemented this in Python (see [pydicom.sr.coding.Code.\_\_eq\_\_](https://github.com/pydicom/pydicom/blob/30cbc79d45cf708fd4afcdd150adc47230d2078e/pydicom/sr/coding.py#L49-L53)).

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+
+    const index = evaluations.findIndex((evaluation) => {
+      const meaning = this.getContentItemCodeMeaning(evaluation);
+      return meaning === codeMeaning;
```
Extracted this method to viewer and updated it

hackermd replied:
We should compare the two items (evaluations) for equality using *CodeValue* and *CodingSchemeDesignator* rather than *CodeMeaning* (see [this comment](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43/files/bb0b14ac99a3ed3bc08dc5da0a5f24ad99ee76aa#r584430833))

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+      this.addOrUpdateEvaluation(feature, evaluation, "Tracking Identifier");
+    }
+
+    if (properties.area) {
```
Extracted this method to viewer and updated it

igoroctaviano commented on src/annotations/markups/freeText.js:
```diff
@@ -0,0 +1,90 @@
+
+const hasMarker = (feature) => !!feature.get("marker");
+
+const FreeTextMarkup = (api) => {
```
Renamed, thanks

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+  addOrUpdateMeasurement(feature, measurement, codeMeaning) {
+    const measurements = feature.get("measurements") || [];
+
+    const index = measurements.findIndex((measurement) => {
```
Logic extracted from this manager and renamed, thanks

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+   * @param {NumContentItem} measurement The measurement content item
+   * @param {string} codeMeaning The code meaning
+   */
+  addOrUpdateMeasurement(feature, measurement, codeMeaning) {
```
Logic extracted from manager and reimplemented, thanks

hackermd replied:
I would suggest passing `code` rather than `codeMeaning` (see [this comment](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43/files/bb0b14ac99a3ed3bc08dc5da0a5f24ad99ee76aa#r584430833))

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+   * @param {TextContentItem} evaluation The feature
+   * @param {string} codeMeaning The code meaning
+   */
+  addOrUpdateEvaluation(feature, evaluation, codeMeaning) {
```
Done, thanks

hackermd replied:
I would suggest passing `code` rather than `codeMeaning` (see [this comment](https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43/files/bb0b14ac99a3ed3bc08dc5da0a5f24ad99ee76aa#r584430833))

igoroctaviano commented on src/annotations/markups/measurement.js:
```diff
@@ -0,0 +1,77 @@
+  let geom = feature.getGeometry();
+  if (geom instanceof LineString) {
+    value = getLength(geom);
+    feature.set("length", value);
```
I have updated this code and moved part of it to utils, the annotation manager only update the feature markup and marker properties based on ROI measurements and evaluations values

igoroctaviano commented on src/annotations/_AnnotationManager.js:
```diff
@@ -0,0 +1,224 @@
+          schemeDesignator: "DCM",
+        }),
+        value: properties.length,
+        unit: unitsSuffix,
```
Updated to coded concept

igoroctaviano commented on src/viewer.js:
```diff
@@ -1152,18 +1547,39 @@ class VolumeImageViewer {
+    console.info(`update ROI ${uid}`);
+
+    const feature = this[_drawingSource].getFeatureById(uid);
+    feature.setProperties(properties);
```
Updated, thanks

igoroctaviano commented on src/viewer.js:
```diff
@@ -520,56 +659,58 @@ class VolumeImageViewer {
-    const baseTotalPixelMatrixColumns = this[_pyramidBaseMetadata].TotalPixelMatrixColumns;
-    const baseTotalPixelMatrixRows = this[_pyramidBaseMetadata].TotalPixelMatrixRows;
+    const baseTotalPixelMatrixColumns = this[_pyramidBaseMetadata]
+      .TotalPixelMatrixColumns;
```
Can you share your vscode config? just to match every change you made to code style

igoroctaviano commented on src/viewer.js:
```diff
@@ -1152,18 +1547,39 @@ class VolumeImageViewer {
+   * @param {string} roi.uid - Unique identifier of the region of interest
+   * @param {object} roi.properties - ROI properties
+   */
+  updateROI({ uid, properties }) {
```
Updated, thanks

hackermd commented:
@igoroctaviano sorry I got very confused because the `File changed` tab showed me the comments in the outdated file context. Please ignore the comments I just made. I will resolve them and start a fresh review.

hackermd reviewed (changes requested):
Thank you @igoroctaviano for addressing the comments and for adding tests and documentation. I have a few additional comments. In addition, I will test the latest changes locally and get back to you on how that goes.

hackermd commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+      schemeDesignator: "DCM",
+    }),
+    value,
+    unit: new dcmjs.sr.coding.CodedConcept({
+      value: unitCodedConceptValue,
+      meaning: unitCodedConceptMeaning,
+      schemeDesignator: "DCM",
+    }),
```
The `MeasurementUnitsSequence` must be a Sequence.

```suggestion
    unit: [new dcmjs.sr.coding.CodedConcept({
      value: unitCodedConceptValue,
      meaning: unitCodedConceptMeaning,
      schemeDesignator: "DCM",
    })],
```

I would actually consider this a bug in dcmjs (see [dcmjs.sr.valueTypes.NumContentItem](https://github.com/dcmjs-org/dcmjs/blob/678543c46df63ec9c11e1f5a97e1776242fea2a6/src/sr/valueTypes.js#L328)). However, I am not sure whether @pieper would be comfortable changing the type of the option and breaking backwards compatibility. Alternatively, we could allow both an array containing one item or an object representing an individual item.

pieper replied:
I really wanted to avoid having to make single-element sequences be Arrays.  My hope was that because the VR of SQ would be enough to handle any ambiguity.  As I recall though there are some corner cases so I think we did try to make the code consistently map SQ to Array.

hackermd replied:
I just thought since its called `unit` rather than `units`, it may be more intuitive to pass an individual item as an object. However, I think we can keep the API as is and just improve documentation.

hackermd replied:
@igoroctaviano this would still need to be addressed. Otherwise the generated SR document will not be standard compliant.

hackermd commented on src/viewer.js:
```diff
@@ -320,32 +398,245 @@ function _coordinateFormatScoord3d2Geometry(coordinates, pyramid) {
+  const unitCodedConceptMeaning = unitSuffixToMeaningMap[unitSuffix];
+
+  if (area) {
+    const nameCodedConceptValue = "Area";
+    const nameCodedConceptMeaning = "42798000";
```
Value and Meaning are switched:

```suggestion
    const nameCodedConceptValue = "42798000";
    const nameCodedConceptMeaning = "Area";
```

igoroctaviano replied:
Fixed on my local env, not sure why it was switched here, maybe old version?

hackermd commented on src/viewer.js:
```diff
@@ -320,32 +398,245 @@ function _coordinateFormatScoord3d2Geometry(coordinates, pyramid) {
+  }
+
+  if (length) {
+    const nameCodedConceptValue = "Length";
+    const nameCodedConceptMeaning = "410668003";
```
Value and Meaning are switched:

```suggestion
    const nameCodedConceptValue = "410668003";
    const nameCodedConceptMeaning = "Length";
```

igoroctaviano replied:
Fixed on my local env, not sure why it was switched here, maybe old version?

hackermd commented on src/viewer.js:
```diff
@@ -320,32 +398,245 @@ function _coordinateFormatScoord3d2Geometry(coordinates, pyramid) {
-  return
+function _updateFeatureMeasurements(map, feature) {
+  const measurements = feature.get(Enums.InternalProperties.Measurements) || [];
+  const area = feature.get(Enums.FeatureMeasurement.Area);
+  const length = feature.get(Enums.FeatureMeasurement.Length);
```
I am wondering whether we should hard-code "Area" and "Length" here.

Could we extend the enum logic?

```js
export const FeatureMeasurement = {
  Area: {
    key: "area",
    code: {CodeValue: "42798000", CodeMeaning: "Area", CodingSchemeDesignator: "SCT"}
  },  
  Length: {
    key: "length",
    code: {CodeValue: "410668003", CodeMeaning: "Length", CodingSchemeDesignator: "SCT"}
  }
};
```

This would allow us to just loop over measurements, extract the measured value from `feature` and construct the SR content item.

hackermd commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+    name: new dcmjs.sr.coding.CodedConcept({
+      value: nameCodedConceptValue,
+      meaning: nameCodedConceptMeaning,
+      schemeDesignator: "DCM",
```
We cannot assume that the coding scheme will always be "DCM". In fact, for both the "Area" and "Length" measurements, the correct coding scheme is "SCT".

I would suggest removing this level of abstract (i.e., removing the `getMeasurementContentItem` utility) and just using `dcmjs.sr.valueTypes.NumContentItem` directly.

hackermd replied:
See [CID 7470 Linear Measurements](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7470.html) and [CID 7471 Area Measurements](http://dicom.nema.org/medical/dicom/current/output/chtml/part16/sect_CID_7471.html)

hackermd commented on src/viewer.js:
```diff
@@ -437,65 +732,77 @@ class VolumeImageViewer {
+    });
+
+    this[_features].on("remove", (e) => {
+      this[_annotationManager].onRemove(feature);
```
The variable `feature` is not defined. I think it should be:

```suggestion
      this[_annotationManager].onRemove(e.element);
```

igoroctaviano replied:
Its fixed on my local env, ill push soon, thanks!

hackermd commented on src/viewer.js:
```diff
@@ -1199,30 +1869,40 @@ class VolumeImageViewer {
   hideROIs() {
-    console.info('hide ROIs')
+    console.info("hide ROIs");
     this[_drawingLayer].setVisible(false);
```
When this method is called the measurements associated with the ROI remain visible.

![screenshot_slim_ROI_show](https://user-images.githubusercontent.com/6855297/113781074-2023aa80-96fe-11eb-93e5-313f602a008a.png)

![screenshot_slim_ROI_hide_bug](https://user-images.githubusercontent.com/6855297/113781067-1c902380-96fe-11eb-8d14-e8f2afe003f6.png)

hackermd replied:
This issue is still persisting. I think we will either need to use the same VectorLayer (`_drawingLayer` stored on viewer object) for ROIs and markups or expose the vector layer(s) created by the markup manager to the viewer so that the `hideROIs()` and `showROIs()` methods can set the visibility of all the vector layers.

hackermd replied:
Probably better to keep them on separate vector layers. That will facilitate toggling markers and markups separately.

hackermd commented on package.json:
```diff
@@ -4,6 +4,7 @@
   "description": "Interactive web-based viewer for DICOM Microscopy Images",
```
Please increase package version.

igoroctaviano commented on src/viewer.js:
```diff
@@ -1117,51 +1753,84 @@ class VolumeImageViewer {
+
+    const feature = this[_drawingSource].getFeatureById(uid);
+
+    _addROIPropertiesToFeature(feature, properties);
```
We do have checks, only specific properties get in

igoroctaviano commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+ * @param {object} contentItem2
+ * @returns {boolean} true if content items equal and false otherwise
+ */
+const isContentItemsEqual = (contentItem1, contentItem2) => {
```
This function uses isCodedConceptEqual which compares the whole thing

hackermd replied:
I think `isCodedConceptEqual` just compares the concept names, i.e., `ConceptNameCodeSequence`, but does not check the values. Therefore, I don't think we can consider the entire content items equal.

hackermd replied:
I would avoid the term "equal" in this context and rename this function to `doContentItemsMatch()` for clarity.

igoroctaviano commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+ * @param {object} codedConcept2
+ * @returns {boolean} true if content items equal and false otherwise
+ */
+const isCodedConceptEqual = (codedConcept1, codedConcept2) => {
```
I'll update, thanks

igoroctaviano commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+const getContentItemNameMeaning = (contentItem) => {
+  const { ConceptNameCodeSequence } = contentItem;
+  return ConceptNameCodeSequence.length
+    ? ConceptNameCodeSequence[0].CodeMeaning
```
I'll do that, thanks

igoroctaviano commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+ * @param {string} unitCodedConceptMeaning the unit coded concept meaning
+ * @returns {object} num content item instance
+ */
+const getMeasurementContentItem = (
```
Sure, thanks

igoroctaviano commented on src/annotations/markups/text.js:
```diff
@@ -0,0 +1,131 @@
+import Fill from "ol/style/Fill";
```
Sure, ill do that, thanks

igoroctaviano commented on src/annotations/markups/measurement.js:
```diff
@@ -0,0 +1,68 @@
+  let value =
+    feature.get(Enums.FeatureMeasurement.Length) ||
+    feature.get(Enums.FeatureMeasurement.Area);
+  let rawValue = Math.round((value / 10) * 100) / 100;
```
Display one, for the actual value i'm using openlayers getArea and getLength functions

igoroctaviano commented on src/annotations/_AnnotationManager.test.js:
```diff
@@ -0,0 +1,200 @@
+  beforeAll(() => {
+    textContentItem = {
+      ConceptNameCodeSequence: {
+        CodeMeaning: "Tracking Identifier",
```
Sure, updating...

hackermd commented on src/utils.js:
```diff
@@ -333,19 +325,208 @@ function mapSlideCoordToPixelCoord(options) {
+ */
+const getContentItemNameCodedConcept = (contentItem) =>
+  contentItem.ConceptNameCodeSequence.length
+    ? contentItem.ConceptNameCodeSequence[0]
```
I would prefer if we would remove that. Workarounds like this one will come back to haunt us one day..

hackermd reviewed (changes requested):
A few remaining issues that I uncovered upon local testing.

hackermd commented on src/annotations/markups/measurement.js:
```diff
@@ -0,0 +1,73 @@
+ * @param {string} units units
+ * @return {string} The formatted measure of this feature
+ */
+export const format = (feature, units, pyramid) => {
```
I noticed two issues with the measurements markup for polygons:

1. The value doesn't seem to be correct (see comparison with resolution bar)
2. The unit is `µm`, but it should be `µm2` (or `mm2`?) in case of area

![Screen Shot 2021-04-30 at 8 24 39 AM](https://user-images.githubusercontent.com/6855297/116695671-eb1e1700-a98e-11eb-960d-c0c7ba482979.png)

igoroctaviano replied:
Good question, i'm not sure if we should use the units of the current resolution or mm for all resolutions.

igoroctaviano replied:
For https://github.com/MGHComputationalPathology/dicom-microscopy-viewer/pull/43#discussion_r623850897, can we create a separate issue? or its a blocker?

igoroctaviano replied:
@hackermd now hide and show should work.

hackermd commented on package.json:
```diff
@@ -30,19 +31,24 @@
-    "mathjs": "^7.1.0",
-    "ol": "^6.3.1"
+    "dcmjs": "^0.18.2",
+    "dicomweb-client": "0.5.2",
+    "mathjs": "7.1.0",
+    "ol": "6.4.3"
```
Let's not specify a specific version for the library, but use `^`.

hackermd commented on package.json:
```diff
@@ -30,19 +31,24 @@
-    "mathjs": "^7.1.0",
-    "ol": "^6.3.1"
+    "dcmjs": "^0.18.2",
+    "dicomweb-client": "0.5.2",
+    "mathjs": "7.1.0",
+    "ol": "6.4.3"
```
```suggestion
    "dicomweb-client": "^0.5.2",
    "mathjs": "^7.1",
    "ol": "^6.5"
```

igoroctaviano replied:
@hackermd updated + tested.

hackermd reviewed (changes requested):
Upgrading the version of Openlayers should fix the issue we have been experiences during local development and testing.

hackermd commented:
@igoroctaviano It is done! Thank you for contributing these excellent new features!
